feat: 로그인 페이지 ui 구현

smb0123 opened the pull request:
### 개요

close #2 

### 작업사항

- 디자인 시스템이 적용된 컴포넌트를 빠르게 머지해야 다른 사용하는 곳에서 적용할 수 있을 것 같아서 로그인 페이지 ui만 구현하였습니다.
- public 폴더의 값은 자주 사용되는 값들이라서 alias 문법을 적용하였습니다.

추후에 새로 이슈를 만들어서 api 로직까지 넣어보겠습니다.
저는 public에 사용되는 값들을 아래와 같이 "이름_이름" 형식으로 저장하고 사용하는 것을 좋아하는 데 다른 분들 의견이 궁금합니다. (이유는 가독성이 좋다고 생각하기 때문입니다.)
```
import Apple_Logo from "@/Apple_Logo.svg";
```

### 체크리스트

- [x] assignees 설정
- [x] label 설정

smb0123 committed:
chore: clsx, tailwind-merge 라이브러리 설치

추가한 이유는 cn 유틸 함수를 만들어서 적용하기 위해서 입니다.

smb0123 committed:
chore: tsconfig 파일에 paths 추가

이유는 public을 경로에 추가함으로써 개발 시간을 단축하기 위함 입니다.

smb0123 committed:
feat: cn 유틸 함수 추가

twMerge, clsx 기능을 하는 함수를 만들어서 편리하기 사용하기
위해 만들었습니다.

smb0123 committed:
feat: globals.css에 색상 추가

디자인 시스템을 적용하기 위해서 색상을 설정했습니다.

smb0123 committed:
feat: Flex 컴포넌트 구현

flex css를 사용하는 컴포넌트를 편리하게
해주기 위해서 Flex 컴포넌트를 구현하였습니다.

smb0123 committed:
feat: Txt 컴포넌트 구현

디자인 시스템을 적용하기 위해서 구현하였습니다.

smb0123 committed:
feat: LoginTitleSection 컴포넌트 구현

로그인 페이지에서 문구로 이미지를 포함하는 컴포넌트를
구현하였습니다.

smb0123 committed:
feat: LoginButtonSection 컴포넌트 구현

로그인 페이지에서 사용하는 카카오, 애플 로그인 ui를
구현하였습니다.

smb0123 committed:
feat: 로그인 페이지 ui 구현

로그인 페이지 UI를 구현하였습니다.

smb0123 commented on src/components/features/pages/LoginButtonSection/index.tsx:
```diff
@@ -0,0 +1,43 @@
+      justifyContent="justify-normal"
+      className="gap-[15px] w-full"
+    >
+      <button className="h-[57px] bg-yellow rounded-lg">
+        <Flex
+          flexDirection="flex-row"
+          alignItems="items-center"
+          justifyContent="justify-center"
+          className="gap-[10px]"
+        >
+          <Image src={Kakao_Logo} alt="카카오 로고" />
+          <Txt fontWeight="font-semibold" fontSize="text-sm">
+            카카오톡으로 3초만에 시작하기
+          </Txt>
+        </Flex>
+      </button>
+      <button className="h-[57px] bg-black rounded-lg">
+        <Flex
+          flexDirection="flex-row"
+          alignItems="items-center"
+          justifyContent="justify-center"
+          className="gap-[10px]"
+        >
+          <Image src={Apple_Logo} alt="애플 로고" />
+          <Txt color="text-white" fontSize="text-sm" fontWeight="font-semibold">
+            Apple로 계속하기
+          </Txt>
+        </Flex>
+      </button>
```
카카오와 애플 로그인으로 컴포넌트를 나누는 작업은 다음 PR에서 하도록 하겠습니다.

geongyu09 commented on src/app/login/page.tsx:
```diff
@@ -0,0 +1,18 @@
+
+export default function LoginPage() {
+  return (
+    <div className="pt-[120px] pb-[120px] pl-[22px] pr-[22px]">
```
해당 부분은 패딩이 없어도 될 것 같아요. 필요하다면 앱에서 주는게 좋을 것 같습니다

추가적으로, 패딩을 굳이 px로 잡지 않아도 됩니다. 오히려 반응형에 좋지 않아요! 
앱을 생각하더라도 px에 너무 사로잡히지는 않아도 되고, 단순히 반응형 웹을 만든다고 생각하도 개발하면 좋을 것 같습니다

smb0123 replied:
패딩 부분은 제거하고 RN에서 작업하는 것으로 알겠습니다!

smb0123 replied:
px은 반응형에 좋지 않는 말이 어떤 의미인가요?? rem 같은 상대적인 단위를 사용하라는 말씀인가용?

geongyu09 commented on src/app/login/page.tsx:
```diff
@@ -0,0 +1,18 @@
+      <Flex
+        flexDirection="flex-col"
+        alignItems="items-center"
+        className="gap-[50px]"
```
gap도 props로 받을 수 있으면 좋을 것 같아요! 자주 사용할 것 같네요!

smb0123 replied:
gap도 props로 받는 것을 생각해봤었는데 gap은 컴포넌트들마다 너무 다를 것 같아서 의미가 없다고 판단했습니다.
그래서 gap을 주고 싶다면 className를 통해서 받는 것을 의도하고 Flex 컴포넌트를 만들었습니다.
그래도 gap을 props로 받는 게 좋아 보이나요?

geongyu09 replied:
흠.. 일단 저에게는 className으로 스타일을 받는 것은 Flex라는 컴포넌트의 특성에서 벗어나서, 정말 불의의 순간에 어쩔 수 없이 사용할 수 있도록 만들어둔 비상구 같은 느낌?인 것 같아요..!
그에 반에서는 gap은 상당히 Flex적인 느낌이 많이 드는 것 같아요! 다른 props들과 동일하게 props로 가져가는게 좋을 것 같다고 생각합니다!

gap이 다양할 것이다라고 판단되는 것은, 아마 number형태로 받으면 되지 않을까 싶어요! 다만, 위처럼 px로 받는 것이 아니라, tailwind의 값으로 설정하면 될 것 같습니다!
tailwind의 값들 (`gap-4` 와 같은 값들)은 결국에 커스텀이 가능해서, 변경에도 용이할 것 같습니다!

smb0123 replied:
gap이라는 props를 number 형으로 입력 받도록 수정하겠습니다!

smb0123 replied:
[4a67770](https://github.com/JNU-econovation/New-Project-FE/pull/5/commits/4a67770609ab4ed20f0f9859d0552fad5a4dc664)
해당 commit에서 작업하였습니다.

geongyu09 commented on src/components/common/shared/layout/Flex/index.tsx:
```diff
@@ -0,0 +1,53 @@
+        justifyContent,
+        alignItems,
+        className,
+        "flex"
```
대체로 공통으로 들어가는 스타일은 가장 위에 놓습니다! 
크게 중요하지는 않을 수 있지만, 수정할 수 있으면 좋을 것 같아요~!

smb0123 replied:
넵 알겠습니다!

geongyu09 commented on src/components/common/shared/ui/Txt/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  fontWeight = "font-normal",
+}: TxtProps) {
+  return (
+    <span className={clsx(color, fontSize, fontWeight, className)}>
```
clsx 대신에 cn을 사용해야할 것 같습니다!

smb0123 replied:
오 이 부분을 놓쳤네요.
감사합니다 👍

geongyu09 commented on src/components/common/shared/ui/Txt/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  | "font-extrabold"
+  | "font-black";
+
+export default function Txt({
```
딴지 거는 것 같을 수 있지만, 정말정말 txt라는 이름이 너무나도 별로인 것 같아요..
txt가 text의 줄인말이겠지만, 왜인지 모르게 확장자 느낌이 너무 강하게 들고, 바로 text라는 느낌을 받지는 못한 것 같아요.
Text로 더 명확하게 이름을 변경하는 것은 어떠신가요?

smb0123 replied:
저도 동의합니다!

geongyu09 commented on src/components/features/pages/LoginButtonSection/index.tsx:
```diff
@@ -0,0 +1,43 @@
+    <Flex
+      flexDirection="flex-col"
+      justifyContent="justify-normal"
+      className="gap-[15px] w-full"
```
width 값도 props로 받을 수 있으면 좋을 것 같아요!
그게 아니라면 fullwidth={boolean} 으로 해도 괜찮을 것 같기도 합니다!

smb0123 replied:
저는 Flex 컴포넌트는 flex에 관한 스타일을 위한 layout이라고 생각해서 이외의 값들은 전부 className으로 받는 게 맞다고 생각합니다. width 받기 시작하면 props가 계속 늘어가게 되게 그렇게 되면 더 이상 해당 컴포넌트는 Flex라는 이름의 컴포넌트로 불리는 게 어려울 것 같다는 생각입니다.

하지만 width까지는 flex에 관한 css가 아니지만 layout이라는 폴더 안에 있으니 허용해줘도 될 것 같기도 하네요.
어떻게 생각하시나요?

seunghyeon77 replied:
Flex 태그의 상위 태그인 div 만들어서 거기에 width 요소 넣는 건 어떠신가요?

smb0123 replied:
그렇게 하면 불필요한 div 태그가 생기기 때문에 Flex의 props로 받는 게 좋을 것 같아요!

geongyu09 commented on src/components/features/pages/LoginTitleSection/index.tsx:
```diff
@@ -0,0 +1,30 @@
+    >
+      <Flex flexDirection="flex-row" alignItems="items-end">
+        <Flex flexDirection="flex-col">
+          <Txt fontWeight="font-bold" className="text-[30px]">
```
나중에는 디자인시스템에 맞추어서 text 크기를 지정할 수 있도록 변경할 수 있으면 좋을 것 같네요!
지금은 글자 크기까지는 시스템화가 안되어있어서 어쩔 수 없지만, 내일 부탁드려보겠습니다

smb0123 replied:
오 저도 해당 부분에 생각을 해봤었는데 디자인 시스템이 폰트에 관련해서 나오더라도 Text 컴포넌트에서만 쓰이기 때문에 전역적으로는 설정을 안 할 것 같습니다. 전역에 설정을 하면 해당 값을 알아내기 위해서 globals.css 파일을 봐야 할텐데 depth가 깊어지면서 개발 시간이 늘어날 것 같습니다.

어떻게 생각하시나요?

geongyu09 commented on src/app/globals.css:
```diff
@@ -1 +1,9 @@
+  --color-sub-gray: #dadada;
+  --color-white: #ffffff;
+  --color-black: #000000;
+  --color-yellow: #fae64d;
```
해당 값은 카카오에서 사용하는 값인 것 같아요!
단순 yellow를 사용하는 것 보다는 더 의미있게 변경해도 좋을 것 같아요! 다른 기존에 사용하던 노란색(tailwind에 있는 노랑)과 혼동되지 않게 kakao-yellow 로 변경하는건 어떤가요?

smb0123 replied:
좋습니다!
기존에 제 생각으로는 kakao라는 이름을 붙이면 해당 색은 카카오가 아닌 다른 곳에서는 사용을 하면 안된다고 생각해서 확장성을 생각해서 추후에 해당 색이 다른 곳에서도 쓰일 수 있으니까 도메인을 붙이지 않은 것인데 지금 생각해보면 과하게 확장성을 생각했던 것 같네요.

geongyu09 commented on src/app/globals.css:
```diff
@@ -1 +1,9 @@
+  --color-main-green: #2b7552;
+  --color-sub-gray: #dadada;
+  --color-white: #ffffff;
+  --color-black: #000000;
```
white, black은 이미 tailwind에 있어서, 제거해도 괜찮을 것 같아요!

smb0123 replied:
오호 알겠습니다.

geongyu09 reviewed (changes requested):
추상화가 정말로 잘 되어있는 느낌을 받았습니다! 정말 다행인 것 같아요!
조금 리뷰가 늦어지긴 했지만, 몇가지 수정을 요청합니다!
조금 오지랖으로 보일 수 있지만, 조금 유난한 팀원이 되어보겠습니다..!

geongyu09 commented:
> 추후에 새로 이슈를 만들어서 api 로직까지 넣어보겠습니다.
저는 public에 사용되는 값들을 아래와 같이 "이름_이름" 형식으로 저장하고 사용하는 것을 좋아하는 데 다른 분들 의견이 궁금합니다. (이유는 가독성이 좋다고 생각하기 때문입니다.)

해당 부분은 저도 동의합니다! 기존에는 카멜 케이스로 사용했었는데 더 가독성이 좋은 것 같아요!

smb0123 commented:
> 추상화가 정말로 잘 되어있는 느낌을 받았습니다! 정말 다행인 것 같아요! 조금 리뷰가 늦어지긴 했지만, 몇가지 수정을 요청합니다! 조금 오지랖으로 보일 수 있지만, 조금 유난한 팀원이 되어보겠습니다..!

저는 이 팀에 들어온 이유 중 하나가 다른 사람의 의견을 들어보고 싶기 때문에 들어 온 것이라서 언제든지 편하게 의견을 줬으면 좋겠습니다.

smb0123 committed:
design: 로그인 페이지 padding 제거

padding은 RN에서 설정하는 게 좋다고 생각했습니다.

smb0123 committed:
rename: Txt 컴포넌트 명 Text로 변경

Txt는 명확한 이름이 아니어서 Text로 이름을
변경하였습니다.

smb0123 committed:
refactor: clsx => cn으로 변경

cn 유틸함수를 사용하기 위해서 변경하였습니다.

smb0123 committed:
chore: cn에 들어가는 공통 스타일 위치 변경

일반적으로 공통으로 들어가는 스타일은 위치를
맨 앞에 두기때문에 수정하였습니다.

smb0123 committed:
chore: yellow에 대한 이름 변경

yellow라는 이름에서 kakao-yellow로 변경하였습니다.

smb0123 committed:
chore: 전역 css 수정

이미 tailwind에 존재하는 white와 black에 대한 설정을
삭제하였습니다.

smb0123 committed:
refactor: Flex 컴포넌트에 gap이라는 props 추가

Flex 컴포넌트에 gap props를 추가했습니다.

geongyu09 commented on src/components/common/shared/layout/Flex/index.tsx:
```diff
@@ -0,0 +1,56 @@
+        flexDirection,
+        justifyContent,
+        alignItems,
+        `gap-[${gap}px]`,
```
개인적으로는.. px보다는 일단 값을 사용하는게 좋을 것 같아요..!
일반 값들은 변경이 가능해서, 수정이 생각보다 편리하고, 무엇보다 px가 아닌 rem을 사용하는 장점이 있어요!
gap-[${gap}] 으로 해도 좋을 것 같아요!

smb0123 replied:
저도 건규님 의견에 동의합니다!
원래는 tailwind 값을 사용하면 저희 디자인에서 사용하는 gap 중에 없는 값도 있었기 때문에 일부러 px을 사용한 것 인데 tailwind는 찾아보니 제공해주는 기본 값이외에도 커스텀이 가능하네요.

geongyu09 commented on src/components/features/pages/LoginButtonSection/index.tsx:
```diff
@@ -0,0 +1,48 @@
+      className="w-full"
+      gap={15}
+    >
+      <button className="h-[57px] bg-kakao-yellow rounded-lg">
```
height를 지정하는 것 보다는 padding으로 크기를 잡는 것이 좋을 것 같습니다!

seunghyeon77 replied:
혹시 height 대신 padding으로 높이를 조정하는 이유가 따로 있나요?
padding으로 조정한다면, 기존 디자인 해놓은 버튼의 크기와 달라질 수 있다고 생각하기 때문입니다!

smb0123 replied:
height와 padding 둘 중 어떤 것을 사용하더라도 크기에는 문제가 없는데 
건규님께서는 padding을 선호하시는 이유가 있을까요?

geongyu09 replied:
반응형으로 개발할 때는 padding으로 개발하는 것이 정답입니다! (반응형이 아니어도 padidng이 맞습니다)

크기를 지정할 때 내부 요소에 맞추어서 ui 가 만들어져야해요. 그렇지 않으면 버튼마다 올바른 height를 지정해줘야합니다. 
만약 버튼 내부에 들어가는 글자나 컴포넌트의 크기가 달라지는데 그때마다 사람이 계산해서 height를 지정하는데는 불필요하게 리소스가 발생합니다. 
추가적으로 글자가 한줄로 되어있다가 화면이 줄어들어 두 줄로 변경되었을 때 화면 크기별로 height를 분기처리해줄수는 없을거에요

height는 컴퓨터가 알아서 계산되도록 만들어주어야 합니다!

geongyu09 commented on src/components/features/pages/LoginButtonSection/index.tsx:
```diff
@@ -0,0 +1,48 @@
+          </Text>
+        </Flex>
+      </button>
+      <button className="h-[57px] bg-black rounded-lg">
```
여기도 px보다는 padding으로 크기를 잡는게 좋을 것 같아요!

geongyu09 reviewed (commented):
정말 수고많으셨습니다! 
웹 앱에서 웹 부분은 px로 하는게 맞을지에 대해서는 잘 모르겠습니다..! 
오히려 px로 인해서 화면 크기가 모두 다른데, 사용자마다 크기가 동일해버리면 문제가 오히려 발생할 수 있을 것 같아요..!
앱을 위한 웹이지만, 단순히 반응형 웹을 만든다는 느낌으로 가면 좋을 것 같아요!
px보다는 rem!

관련해서, gap 부분도 px를 사용하는 것 보다는 rem을 적극 사용할 수 있는 방안으로 변경하면 좋을 것 같아요! 
(일반 tailwind 값 사용하기)

관련해서 질문이나 생각이 있다면, 남겨주시면 감사할 것 같아요..!
gap을 수정하는 것을 바라지만.. 혹시나 px을 사용해야하는 이유가 있다면 말씀 해주세요!

smb0123 commented:
> 정말 수고많으셨습니다! 웹 앱에서 웹 부분은 px로 하는게 맞을지에 대해서는 잘 모르겠습니다..! 오히려 px로 인해서 화면 크기가 모두 다른데, 사용자마다 크기가 동일해버리면 문제가 오히려 발생할 수 있을 것 같아요..! 앱을 위한 웹이지만, 단순히 반응형 웹을 만든다는 느낌으로 가면 좋을 것 같아요! px보다는 rem!

@geongyu09  제가 알기로는 rem은 상대 단위이긴 하지만 html 태그의 font-size를 기준으로 하는 것으로 알고 있습니다.
따라서 rem으로 하더라도 사용자마다 동일한 크기로 보인다고 알고 있어요. 그리고 웹 기준으로는 브라우저의 font-size를 변경해야만 사용자마다 다르게 보여서 px을 사용하나 rem을 사용하나 똑같다고 생각합니다. (물론 브라우저의 font-size를 변경한다면 말이 달라지는 합니다. 이 경우를 말씀하신 거라면 rem으로 바꾸는 게 맞다고 생각하지만 사용자마다 화면의 크기가 다른 데 px을 사용하면 모든 사용자에게 동일한 크기가 보여서 rem으로 변경하는 게 좋다는 것은 잘 모르겠습니다.)
혹시 제가 잘못 알고 있는 오개념이 있을까요?? 아니면 앱에서는 다르게 동작하는 건지 궁금합니다.

smb0123 reviewed (commented):
@geongyu09 그리고 제가 작성한 코드에서 px을 사용한 부분들은 테일윈드에서 제공해주지 않는 값을 사용하기 위해서 px을 사용한 것인데 rem으로 수정하도록 하겠습니다.

smb0123 committed:
refactor: Flex 컴포넌트 수정

tailwind는 동적인 className을 인식하지 못해서 gap을 string으로 전부
받아오는 방식으로 변경하였습니다. 그리고 기본 값을 w-full로 설정하고
maxWidth를 옵셔널로 받도록 변경하였습.

geongyu09 commented:
반응형 웹을 만들기 위해서 px가 아닌 rem을 사용하는 것은 알려진 사실입니다! 
rem을 사용해야 하는 이유에 대해서 몇 가지 생각을 남겨보겠습니다.

1. 다양한 디바이스 대응: 컴퓨터 브라우저와 모바일 기기의 화면 표시 방식이 다릅니다. 또한 스마트폰 종류에 따라 화면 크기가 달라질 수 있어요. (갤럭시 플립과 같은 폴더블 기기는 실제로 표시 크기가 다를 수 있습니다!) rem을 사용하면 이러한 다양한 화면 환경에 유연하게 대응할 수 있습니다.
2. 해상도 차이 대응: 정말 다양한 화면의 크기마다 해상도가 다를 수 있어요! 화면이 커도 해상도가 낮으면 글자가 커 보일 수 있습니다. rem을 사용하면 이에 유연하게 대응하기가 편해집니다. (rem 크기만 변경하면 그 크기에 맞춰서 모든 화면이 조정됩니다. px는 하드코딩 느낌이라면 rem은 전역변수 하나를 두고 쉽게 전체적으로 변경이 가능합니다!)
3. 일관된 비율 유지: rem을 사용하면 디자인 전체에서 요소 간의 비율이 일관되게 유지됩니다 
4. 유지보수 용이성: 3번, 2번과 연장선상으로, 디자인 변경이 필요할 때 루트 요소의 폰트 크기만 조정하면 전체 레이아웃이 비례적으로 변경됩니다.  유지보수에 정말 좋습니다

smb0123 committed:
chore: font-size 변경

px 대신 tailwind에서 제공하는 기본 값으로 변경하였습니다.

smb0123 committed:
chore: 버튼의 크기를 padding으로 설정

버튼의 크기를 height가 아니라 padding으로 변경하였습니다.

smb0123 reviewed (commented):
말씀하신 변경사항들은 전부 반영하였는데 아래와 같은 의문점이 발생하였습니다.

tailwind는 동적인 스타일을 무시해서 gap과 width를 숫자로 입력하면 스타일을 적용하는 과정에서 인식을 못합니다. 따라서 string으로 tailwind에서 입력하는 값 그대로를 입력 받도록 하였는데 string으로 하면 gap, maxWidth props를 사용이 불편할 것이라고 생각하지만 유니온 타입으로 만들자고 하니 tailwind에서 제공해주는 값을 다 입력하는 것은 아니라고 생각을 해서 일단은 string으로 했는데 유니온 타입으로 하고 사용하는 값이 생길 때마다 Flex 컴포넌트를 수정하는 방향으로 가는게 좋을까요?

geongyu09 commented on src/components/common/shared/layout/Flex/index.tsx:
```diff
@@ -0,0 +1,57 @@
+        "flex",
+        "w-full",
+        maxWidth,
+        gap,
```
global.css에 아래의 줄을 추가하면 잘 되는 것 같습니다!

`@source inline("gap-{{0..12..1},14,{16..64..4},{72..96..8}}");`

ref: https://tailwindcss.com/docs/detecting-classes-in-source-files#safelisting-specific-utilities

geongyu09 replied:
변경한다면 아래처럼 변경하면 될 것 같아요~!

```
import { cn } from "@/utils/cn";

interface FlexProps {
//...
  gap?: number;
}

// ...

export default function Flex({
// ...
  gap,
}: FlexProps) {
  return (
    <div
      className={cn(
        //...
        gap ? `gap-${gap}` : ""
      )}
    >
      {children}
    </div>
  );
}

```

geongyu09 commented on src/app/login/page.tsx:
```diff
@@ -0,0 +1,12 @@
+
+export default function LoginPage() {
+  return (
+    <Flex flexDirection="flex-col" alignItems="items-center" gap="gap-12">
```
단순히 gap을 주기위한 용도라면( flex 포지셔닝과는 조금 거리감이 있는, 두 컴포넌트의 간격을 주는 경우), display를 변경하는 flex 보다는 Spacing과 같은 컴포넌트를 사용하는게 더 좋을 것 같습니다!
해당 컴포넌트는 @seunghyeon77 님이 개발중이므로, Spacing 코드까지 dev 브랜치에 머지되면 그때 변경해도 좋을 것 같습니다!

지금은 수정 안해도 괜찮을 것 같아요

geongyu09 replied:
변경하게 된다면 아래의 내용처럼 바뀌겠네요!
```
    <>
      <LoginTitleSection />
      <Spacing gap={12} />
      <LoginButtonSection />
    </>
```

seunghyeon77 replied:
엇 저 또한 gap을 줄 때 민보님이 작성하신 방법으로, Flex 컴포넌트에 gap을 주는 형식으로 진행 하였는데, 수정해야될까요?
그리고 혹시 gap이라는 css 요소가 flex랑 같이 쓰이는 걸로 알고 있는데, 그럼 저기서의 gap은 이름만 gap인 것인가요? 아니라면 단독으로 쓰일 수 있는지 궁금해요!

geongyu09 replied:
> 엇 저 또한 gap을 줄 때 민보님이 작성하신 방법으로, Flex 컴포넌트에 gap을 주는 형식으로 진행 하였는데, 수정해야될까요?

음..! 상황마다 다를 것 같아요
단순히 두 독립적인 컴포넌트에 특별한 포지션을 가지도록 만드는 것이 아니라면 굳이 flex를 사용할 필요가 없죠. 
flex는 자식 요소들의 위치를 잡을 때 사용합니다. 기존이 block 형식을 가져도 가로로 놓입니다. 그리고 width가 일반적으로 full인 것이 아니게 바뀝니다.

그렇다면 과연 하단 컴포넌트 두개를 띄우기 위해서 flex를 사용해야할까요? 
flex를 사용하게 되면 의도하지 않게 하위 요소들이 차지하는 범위가 달라질 수 있어요. 
이를 변경하려고 flex의 기본 속성이 아닌 일반 div의 속성처럼 상하로 두개 만들기 위해서 direction을 또 바꿔주어야 합니다. 
즉 일을 중복하게 되었네요. 
원래 div는 상하로 놓이는데 간격을 주기 위해서 flex로 두어서 좌우에 두게 하고, 이를 상하로 변경하였으니까요.!

이런 접근보다는 maring-top을 LoginButtonSection 에 주면 되지 않을까? 라는 생각을 해보면 좋을 것 같아요.
그런데 컴포넌트 자체가 maring을 가지고 있으면 당연하게 재사용성이 떨어지겠죠?
그래서 이를 따로 상위에서 주어야 합니다. 
그렇다면, 이렇게 margin-top을 주는 경우가 이번 한번 뿐일까요? 아니겠죠..!
이렇게 각 컴포넌트들의 간격을 줄 수 있는 레이아웃만 가지는 컴포넌트를 하나 만들어두면 두고두고 사용할 수 있겠죠? 
실제로도 이렇게 사용하기도 하고요!

이러한 부분에 대해서 "엇 나도 그런것 같은데, 수정할까요?" 라는 것 보다는 조금더 무게감 있게 먼저 깊이있는 고민을 해보면 좋을 것 같습니다.
왜 이 사람이 이런 생각을 했는지를 끊임없이 고민할 수 있으면 좋을 것 같아요.!
그리고 본인의 주장을 정하고 그에대한 근거를 최대한 생각해볼 수 있으면 좋을 것 같아요!
그리고 그 이후에 나는 이런 결론이 났는데, 왜 이렇게 생각하였나요? 라는 질문을 해주면 더 좋은 팀 문화가 만들어질 것 같아요..!
"수정해야할까요?" 라는 말 보다는 "저는 ~한 부분때문에 flex를 사용하고자 했습니다. 이러한 곳에서는 flex가 더 잘 맞는 것 같은데, 어떻게 생각하시나요?" 형태를 기대했었습니다!

제 커맨트 내에서 생각해볼만한 거리를 남겨보겠습니다. 한번씩 생각해보면 좋을 것 같아요
- flex는 언제 사용을 고려해야할까?
- 왜 예시를 들때 padding도 있는데 margin을 사용하는 걸까?
- margin도 margin-bottom을 줄 수 있는데 왜 margin-top을 쓴다고 하였을까? 

작성하다보니 조금 많은 쓴소리를 하게 되었는데.. 제가 지금 너무 예민한가봅니다. 적당히 걸러들어주세요...🥲

아래는 참고할만한 레퍼런스입니다! 
토스 slash : https://www.slash.page/ko/libraries/react/emotion-utils/src/spacing.i18n
카테캠 당근 개발자 작성 코드 : https://github.com/geongyu09/react-deploy/blob/geongyu09-step3/src/components/common/layouts/Spacing/index.tsx

geongyu09 replied:
> 혹시 gap이라는 css 요소가 flex랑 같이 쓰이는 걸로 알고 있는데, 그럼 저기서의 gap은 이름만 gap인 것인가요? 아니라면 단독으로 쓰일 수 있는지 궁금해요!

props에 대한 이름은 편하게 정해주셔도 됩니다. 
flex에서 사용하는 gap과 이름이 겹치므로 변경해주셔도 됩니다. 
토스에서는 size 로 사용하네요!

구현은.. 어떻게 했을까요? 이는 @seunghyeon77 님의 몫입니다

seunghyeon77 replied:
저는 블록 단위의 요소들의 여백을 한꺼번에 주기 위해, gap을 사용하였습니다!

텍스트와 input, button 태그를 하나의 블록으로 만들고 Spacing 컴포넌트를 통해서 여백을 각각 주는 것보다, 블록 (div)을 줄이고 여백도 한꺼번에 주기 위해 Flex 박스를 사용했습니다! 

하지만 Flex를 준다면, 브라우저가 레이아웃을 계산하는 것이므로 필요한 경우가 아니라면 사용을 줄여야겠습니다.

모든 요소들이 경우에 따라 다른데도 불구하고, 제가 생각을 깊이 있게 하지 않고 말했네요. 죄송합니다!

다음부터는 고민해보고 저의 주장을 펼치도록 하겠습니다..!

그리고 해당 고민 거리는 천천히 생각해보고 의논 해볼 수 있도록 하겠습니다. 감사합니다 :)

geongyu09 reviewed (commented):
수고 많으셨습니다! 
gap 부분은 리뷰에도 적었지만, 단순히 string으로 받는 거라면 classname을 받는 거와 별다른게 없어지게 되어서 좋은 방법이 아닐 것 같습니다..!
빌드시에 정적으로 모든 값을 만들어줄 수 있어야 할 것 같아요!

방법은 커맨트에 남겼습니다!
해당 내용만 수정되면 될 것 같아요!

smb0123 committed:
refactor: Flex 컴포넌트의 gap props의 타입을 number로 변경

globals.css에 gap의 다양한 형태를 미리 생성해두고 이를 사용하도록
변경하였습니다. 기존의 string 타입으로 진행하게 되면 gap을 props로 받는
의도가 명확해지지 않기 때문에 number 타입으로 변경하였습니다.

smb0123 reviewed (commented):
Flex 컴포넌트의 gap 타입을 number로 변경하였습니다.
건규님께서 approve 해주시면 머지하도록 하겠습니다.
승현님께서 작업 중인 컴포넌트를 적용해서 머지하는 것도 좋지만 제가 해당 작업 내용을 가지고 바로 다음 작업을 시작해야해서 먼저 머지해야 할 것 같습니다.

geongyu09 reviewed (approved):
고생많으셨습니다 👍👍👍👍

seunghyeon77 commented:
제가 Button ui 컴포넌트와 Spacing ui 컴포넌트 만들어서 적용시키겠습니다!
